services/api/src/api/routes/rows.py

```diff
@@ -0,0 +1,319 @@
+            with StepProfiler(method="rows.index", step="get the dataset's features"):
+                features = Features.from_arrow_schema(parquet_files[0].schema.to_arrow_schema())
+
+            supported_columns, unsupported_columns = [], []
+            for column, feature in features.items():
+                str_feature = str(feature)
+                str_column = str(column)
+                if any(magic_string in str_feature for magic_string in UNSUPPORTED_FEATURES_MAGIC_STRINGS):
+                    unsupported_columns.append(str_column)
+                else:
+                    supported_columns.append(str_column)
+
+            with StepProfiler(method="rows.index", step="create the row group offsets"):
+                row_group_offsets = np.cumsum(
+                    [
+                        parquet_file.metadata.row_group(group_id).num_rows
+                        for parquet_file in parquet_files
+                        for group_id in range(parquet_file.metadata.num_row_groups)
+                    ]
+                )
+            with StepProfiler(method="rows.index", step="create the row group readers"):
+                row_group_readers = [
+                    partial(parquet_file.read_row_group, i=group_id, columns=supported_columns)
+                    for parquet_file in parquet_files
+                    for group_id in range(parquet_file.metadata.num_row_groups)
+                ]
+            return row_group_offsets, row_group_readers, features, unsupported_columns
+
+    return index
+
+
+def query(offset: int, length: int, row_group_offsets: Any, row_group_readers: List[RowGroupReader]) -> pa.Table:
+    """Query the parquet files
+
+    Note that this implementation will always read one row group, to get the list of columns and always have the same
+    schema, even if the requested rows are invalid (out of range).
+
+    Args:
+        offset (int): The first row to read.
+        length (int): The number of rows to read.
+        row_group_offsets (Any): The row group offsets. See index().
+        row_group_readers (List[RowGroupReader]): The row group readers. See index().
+
+    Returns:
+        pa.Table: The requested rows.
+    """
+    if (len(row_group_offsets) == 0) or (len(row_group_readers) == 0):
+        raise ParquetResponseEmptyError("No parquet files found.")
+    last_row_in_parquet = row_group_offsets[-1] - 1
+    first_row = min(offset, last_row_in_parquet)
+    last_row = min(offset, offset + length - 1, last_row_in_parquet)
+    first_row_group_id, last_row_group_id = np.searchsorted(row_group_offsets, [first_row, last_row], side="right")
+    pa_table = pa.concat_tables([row_group_readers[i]() for i in range(first_row_group_id, last_row_group_id + 1)])
+    first_row_in_pa_table = row_group_offsets[first_row_group_id - 1] if first_row_group_id > 0 else 0
+    return pa_table.slice(offset - first_row_in_pa_table, length)
+
+
+class FeatureItem(TypedDict):
```

Comment:
nit: Maybe we could move this to libcommon since it is used in first-rows worker  and will be used in first-rows-from-parquet as well

Reply:
Yes, I agree. We already discussed this with @lhoestq, and decided to first get /rows working, and then factorize the common code between /rows and /first-rows. Is it OK for you?